cmd/kube-proxy/app/server_others.go

```diff
@@ -145,6 +152,38 @@ func newProxyServer(
 		return nil, fmt.Errorf("cannot determine detect-local-mode: %v", err)
 	}
 
+	var nodeInfo *v1.Node
+	if detectLocalMode == proxyconfigapi.LocalModeNodeCIDR {
+		klog.Infof("Watching for node %s, awaiting podCIDR allocation", hostname)
+
+		// since allocators can assign the podCIDR after the node registers, we do a watch here to wait
```

Comment:
sorry if is a dumb question, I'm not very familiar with the K8S API `Watch()` implementation.
What happens if the podCIDR was already assigned to the node? 
should we check first if podCIDR exists and, if not, add the watcher?

Reply:
that works. watch first returns the current state of the object and then any changes. Tested by restarting kube-proxy and it works correctly.

Reply:
direct use of Watch invocations needs to correctly handle dropped/expired watches. Does this do that?

Also see https://github.com/kubernetes/kubernetes/issues/74022 for issues with watches started without specifying a resourceVersion

Is there a reason this isn't using one of the watch tools helpers (like `k8s.io/client-go/tools/watch#NewRetryWatcher`) that manages restarting the watch for you?

Reply:
main reason was I did not know? :-)

Will take a look at that.

Reply:
UntilWithSync might be the simplest option (you can pass in a context with a timeout to control how long you wait)

Reply:
Thanks. But it looks like it works with  cache.ListerWatcher and that needs to be created first and then passed to it. It does not work directly with the Watch interface?

Reply:
A watcher is not arbitrarily restartable. To handle restart for you, it needs both the lister and the watcher. You can make a lister/watcher in place like this (passing in the same context with a deadline you pass to UntilWithSync):

```go
fieldSelector := fields.OneTermEqualSelector("metadata.name", nodeName).String()
lw := &cache.ListWatch{
	ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
		options.FieldSelector = fieldSelector
		return client.CoreV1().Nodes().List(ctx, options)
	},
	WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
		options.FieldSelector = fieldSelector
		return client.CoreV1().Nodes().Watch(ctx, options)
	},
}
```

Reply:
Thank you. This is really helpful. I was looking to see how much of the cache interface I'll have to use. This answered a lot of those questions. Thanks!

Reply:
Updated.